src/main/java/htsjdk/samtools/BinaryBAMShardIndexWriter.java

```diff
@@ -0,0 +1,158 @@
+      }
+
+      // write metadata "bin" and chunks        
+      if (metaData != null)
+          writeChunkMetaData(metaData);
+
+      // write linear index
+
+      final LinearIndex linearIndex = content.getLinearIndex();
+      final long[] entries = linearIndex == null ? null : linearIndex.getIndexEntries();
+      final int indexStart = linearIndex == null ? 0 : linearIndex.getIndexStart();
+      final int n_intv = entries == null ? indexStart : entries.length + indexStart;
+      codec.writeInt(n_intv);
+      if (entries == null) {
+          return;
+      }
+      // since indexStart is usually 0, this is usually a no-op
+      for (int i = 0; i < indexStart; i++) {
+          codec.writeLong(0);
+      }
+      for (int k = 0; k < entries.length; k++) {
+          codec.writeLong(entries[k]);
+      }
+      try {
+          codec.getOutputStream().flush();
+      } catch (final IOException e) {
+          throw new SAMException("IOException in BinaryBAMIndexWriter reference " + content.getReferenceSequence(), e);
+      }
+  }
+
+  /**
+   * Writes out the count of records without coordinates
+   *
+   * @param count
+   */
+  @Override
+  public void writeNoCoordinateRecordCount(final Long count) {
+      codec.writeLong(count == null ? 0 : count);
+  }
+
+  /**
+   * Any necessary processing at the end of the file
+   */
+  @Override
+  public void close() {
+      codec.close();
+  }
+
+  private void writeBin(final Bin bin) {
+      final int binNumber = bin.getBinNumber();
+      if (binNumber >= GenomicIndexUtil.MAX_BINS){
+          throw new SAMException("Unexpected bin number when writing bam index " + binNumber);
+      }
+      
+      codec.writeInt(binNumber);
+      if (bin.getChunkList() == null){
+          codec.writeInt(0);
+          return;
+      }
+      final List<Chunk> chunkList = bin.getChunkList();
+      final int n_chunk = chunkList.size();
+      codec.writeInt(n_chunk);
+      for (final Chunk c : chunkList) {
+          codec.writeLong(c.getChunkStart());
+          codec.writeLong(c.getChunkEnd());
+      }
+  }
+
+  /**
+   * Write the meta data represented by the chunkLists associated with bin MAX_BINS 37450
+   *
+   * @param metaData information describing numAligned records, numUnAligned, etc
+   */
+  private void writeChunkMetaData(final BAMIndexMetaData metaData) {
+      codec.writeInt(GenomicIndexUtil.MAX_BINS);
+      final int nChunk = 2;
+      codec.writeInt(nChunk);
+      codec.writeLong(metaData.getFirstOffset());
+      codec.writeLong(metaData.getLastOffset());
+      codec.writeLong(metaData.getAlignedRecordCount());
+      codec.writeLong(metaData.getUnalignedRecordCount());
+
+  }
+
+  private void writeHeader(int nRef) {
+      // magic string
+      final byte[] magic = BAMFileConstants.BAM_INDEX_MAGIC;
+      codec.writeBytes(magic);
+      codec.writeInt(nRef);
+  }
+
+  private void writeNullContent() {
+      codec.writeLong(0);  // 0 bins , 0 intv
```

Comment:
Does `intv` mean interval?  It might be nice to expand on the comment here and why this is necessary, even if folks reading through the code will eventually figure it out.